# The RNA-Seq Toolkit #

Development of the RNA-Seq Toolkit is led by Scott Givan. Contributions have also been made by colleagues at the Informatics Research Core Facility at the University of Missouri and the Bioinformatics and Biostatistics Core at the Van Andel Research Institute.
The RNA-Seq Tookit is under constant development and improvement. If you have any questions or concerns, please contact Scott at sgivan@yahoo.com.

The current default branch, FirstPBS, uses [STAR](https://github.com/alexdobin/STAR) to align RNA-Seq reads to a reference genome and [DESeq2](https://www.bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html) to run a differential expression analysis. It is also adpated to use the [PBS](https://www.pbsworks.com) resource allocation system.

For a full tutorial, see [the RNA-Seq-Toolkit wiki](https://github.com/sgivan/RNA-Seq-Toolkit/wiki/0.-How-to-Use-the-RNA-Seq-Toolkit).

Also, if you use the RNA-Seq-Toolkit, please cite our [publication](http://www.springerlink.com/content/mu65310036466275/)

@sgivan
