---
title: "DESeq2 DE Analysis"
output:
    html_document:
        toc: true
        toc_depth: 2
        keep_md: true
        fic_caption: true
        extension: pipe_tables
---
```{r load libraries, message=FALSE, warning=FALSE }

library(tidyverse)
library(edgeR)
library(DESeq2)
library($orgdb)
library(vegan)
library(pheatmap)
library(gprofiler2)
```

```{r create plot_Volcano subroutine, include=FALSE}
# Thanks Stephen Turner (see https://gist.github.com/stephenturner/4a599dbf120f380d38e7#file-volcanoplot-r)
plot_Volcano <- function(DE_results_filename,labelPoints=FALSE) {

#    res <- read.table(DE_results_filename, na.strings="", sep="\t",header=TRUE)
    res <- read_tsv(DE_results_filename, col_names=T)
#    row.names(res) <- res$$Gene

    # No NA's in log2FoldChange, right?
    # True is also 1, right?
    log2fc <- res$$log2FoldChange
    circle_sizes <- rep(1,length(log2fc))

    res <- res%>%mutate(color = ifelse(abs(log2FoldChange) >= 1,ifelse(padj<=0.05 , "green", "orange"),ifelse(padj<=0.05,"red","black")))
    max_y <- max(-log10(res$$pvalue)) + 0.5

    ggplot(res, aes(x=log2FoldChange, y=-log10(pvalue), ymax=max_y, ymin=0)) +
    geom_point(aes(colour=color), size=2.5) +
    scale_colour_manual(name="", values = c("green"="green", "orange"="orange", "red"="red", "black"="black")) +

    theme(axis.text=element_text(size=12), axis.title=element_text(size=16)) + # Make axis text and title bigger
    guides(color=FALSE)

}
```

## Create the merged data object
```{r merge data, message=FALSE, warning=FALSE}
merged.data <- read_tsv(paste0("../","$aligndir","/Sample_1/PE_ReadsPerGene.out.tab"), col_names=F, skip=4)[,c(1,$strand)]
sampnames <- c("GENEID", "Sample_1")
eod <- $cntCont + $cntExp
for (i in 2:eod) {
    sname <- paste0("Sample_",i)
    fpth <- paste0("../", "$aligndir","/", sname, "/PE_ReadsPerGene.out.tab")
#    sampdata <- read.delim(fpth, header=T, row.names=NULL, sep="\t", stringsAsFactors=F)
    sampnames <- append(sampnames, sname)
    sampdata <- read_tsv(fpth, col_names=F, skip=4)[,c(1,$strand)]
    merged.data <- merge(merged.data, sampdata, by="X1")
}
colnames(merged.data) <- sampnames
```
## Create gene expression data matrix

The gene_cnt_matrix.tab file, loaded in the next command, should have gene IDs in the first column.
Then, each sample should occupy its own column.
Control coumns should be the columns after the first.
Experimental columns should follow after all the control columns.
the header should reflect the columns. For example:
GeneID    Sample_1    Sample_2    Sample_3    Sample_4

```{r create data matrix, message=FALSE, warning=FALSE}
data <- tibble::column_to_rownames(merged.data, var="GENEID")

rnaseqMatrix = data

# Keep only rows with at least two cpms greater than one
rnaseqMatrix = rnaseqMatrix[rowSums(cpm(rnaseqMatrix) > 1) >= 2,]
```

## Create a dataframe of condition labels
```{r create conditions df, message=FALSE, warning=FALSE}
condition = data.frame(condition=factor(c(rep("Cont", $cntCont), rep("Exp",$cntExp))))
sample_IDs <- colnames(rnaseqMatrix)

# Add rownames that correspond to sample IDs
rownames(condition) = sample_IDs
```

## DESeq2: Calculate differential expression using count data
```{r calc DESeq2 DE}
ddsFullCountTable <- DESeqDataSetFromMatrix(
    countData = rnaseqMatrix,
    colData = condition,
    design = ~ condition)
dds = DESeq(ddsFullCountTable)
```
## Calcluate statistical significance of clustering in PCA analysis
```{r adonis}
vsd <- vst(dds, blind=F)
vsd.df.t <- t(as.data.frame(assay(vsd)))
vsd.adonis <- adonis(vsd.df.t ~ colData(dds)$$condition, method="eu", permutations=9999)
vsd.adonis
```

## Plot PCA Reslts
```{r plot PCA results, echo=FALSE, message=FALSE, fig.cap=pcaption, results=FALSE}
# plotPCA can't take dds straight
rldds <- rlog(dds)

pcaption <- paste0("A Principal Components Analysis (PCA) plot is a way to visualize the overall relationship between sample replicates based on the empirical data. In ideal situations, the points in the plot representing individual replicates for a given sample type should cluster together more closely than to the points representing a different sample type. The PERMANOVA analysis is a common statistical test to estimate whether the clustering of samples is greater than would be expected by chance. A common threshold of significance is p < 0.05. The p value for this analysis is p = ", vsd.adonis$$aov.tab$$`Pr(>F)`[1])

plotPCA(rldds,intgroup=c('condition')) # This will give groups, but not sample names within the group

svg(paste0("$prefix", "_PCA.svg"))
plotPCA(rldds,intgroup=c('condition')) # This will give groups, but not sample names within the group
dev.off()

```

[svg PCA plot](./DESeq2_PCA.svg)

## Set up contrast and generate DE results
```{r set up contrast, message=FALSE, warning=FALSE}
contrast=c("condition","Exp","Cont")
res = results(dds, contrast)
dds.res <- res

res = cbind(experimental="Exp", control="Cont", as.data.frame(res))
```

## Add baseMean Exp and Control
```{r add baseMeans, message=FALSE, warning=FALSE}
baseMeanExp     <- rowMeans(counts(dds, normalized=TRUE)[,colData(dds)$$condition == "Exp"])
baseMeanControl <- rowMeans(counts(dds, normalized=TRUE)[,colData(dds)$$condition == "Cont"])
res = cbind(baseMeanExp, baseMeanControl, as.data.frame(res))

res$$padj[is.na(res$$padj)]  <- 1
```

## Map gene ID's to common gene names
wrap the call to mapIds() in a tryCatch() routine to catch errors
if mapIds fails, use gprofiler2::gconvert()

```{r get gene symbols}
geneSymbol <- row.names(res)

rslt <- tryCatch(
                    expr = {
                        geneSymbol <- mapIds($orgdb, keys=row.names(res), column="SYMBOL", keytype="$dbkey", multiVals="first")
                    },
                    error = function(e){
                        message("mapping gene symbols with org DB generated an error")
                        return(e)
                    },
                    warning = function(w){
                        message("mapping gene symbols with org DB generated a warning")
                        print(w)
                    },
                    finally = {
                        message("moving on")
                    }
                 )

#
# if mapIds() generated an error, try to get gene names using the gprofiler2 function gconvert
if (inherits(rslt, 'simpleError')) {
    message("because of error using org DB, now try gprofiler2 to get gene symbols")
# sometimes this fails to, so wrap in tryCatch()
    rslt <- tryCatch(
                     expr = {
                        gcrslt <- gconvert(row.names(res), organism="$gProfilerkey")
                        geneSymbol <- gcrslt$$name
                     },
                     error = function(e){
                         message("gconvert also failed. Will create report without common gene names")
                         return(e)
                     },
                     warning = function(w){
                         message("gconvert generated a warning")
                         return(w)
                     },
                     finally = {
                         message("moving on")
                     }
                     )
}

#
# let's assume that if we've got this far that geneSymbol contains gene names
bestGeneDescriptor <- gsub("'","_",geneSymbol)
bestGeneDescriptor[ is.na(bestGeneDescriptor) ] <- row.names(res)[ is.na(bestGeneDescriptor) ]

res <- cbind(GeneSymbol=geneSymbol, as.data.frame(res))
res <- cbind(Gene=row.names(res), as.data.frame(res))
res <- cbind(BestGeneDescriptor=bestGeneDescriptor, as.data.frame(res))
```

## Write DE results and count matrix to separate files

```{r write DE results to file, message=FALSE, warning=FALSE}
write.table(as.data.frame(res[order(res$$pvalue),]), file=paste0("$prefix", "_DESeq2_DE_results.txt"), sep="\t", na="", row.names=FALSE,quote=FALSE)
write.table(rnaseqMatrix, file=paste0("$prefix", "_DESeq2_count_matrix.txt"), sep="\t", na="", row.names=FALSE,quote=FALSE)
```
The DE results will be in tab-delimited text files that can be opened by Microsoft Excel or imported into a spreadsheet. You can download the files
using the links, below. The columns in the DE_results file are as indicated in the following table.

| Column Header | Description of Column |
|:-------------:|:----------------------|
| BestGeneDescriptor | Gene _Symbol_, if it exists. Otherwise, a database ID. |
| Gene | Database ID; ie ENSMUSG00000026640. |
| GeneSymbol | The accepted Gene Symbol; ie, Rac1. |
| baseMeanExp | Normalized expression level of the gene in the experimental sample type, expressed as counts. |
| baseMeanControl | Normalized expression level of the gene in the control sample type, expressed as counts. |
| experimental | The name of the experimental sample type. |
| control | The name of the control sample type. |
| baseMean | The average normalized expression level of the gene across both sample types. |
| log2FoldChange | The log2-transformed ratio of (baseMeanExp)/(baseMeanControl). |
| lfcSE | The standard error associated with the calculation of the log2 ratio. |
| stat | An internal statistic used by [DESeq2](https://bioconductor.org/packages/release/bioc/html/DESeq2.html). |
| pvalue | The p-value associated with the log2FoldChange occurring by chance within this data set. |
| padj | The False Discovery Rate corrected p-value. |

### Download Links

[DE results (tab-delimited text)](./DESeq2_DESeq2_DE_results.txt)
[count matrix (tab-delimited text)](./DESeq2_DESeq2_count_matrix.txt)


## Generate volcano plots
```{r create volcano plots, echo=FALSE, fig.cap=volcaption, message=FALSE, results=FALSE}

volcaption <- paste("A Volcano plot is a way to visualize the distribution of expression ratios (X-axis) and p-values (Y-axis) for a single data set. \
It is a summary plot, so not a lot of detail can be inferred. However, you should see a relatively broad distribution of points across the plot and \
the lowest p-values (plotted higher on the Y-axis) should spread fairly evenly across a range of expression ratios. In the plot, orange points \
represent expression ratios greater than 2X (either up or down), red points represent genes whose FDR-corrected p-values are < 0.05 and green points \
are genes where both properties occur (fold-change > +/- 2X and FDR < 0.05).")

plot_Volcano(paste0("$prefix", "_DESeq2_DE_results.txt"))

# svg file
svg(paste0("$prefix", "_Volcano.svg"))
plot_Volcano(paste0("$prefix", "_DESeq2_DE_results.txt"))
dev.off()

```

[svg volcano plot](./DESeq2_Volcano.svg)

## Create heatmap plots
```{r heatmap, echo=FALSE, fig.cap=hmcap}

hmcap <- paste("A Heatmap is a way to visualize the expression of individual genes across sample types. Each column represents a sample replicate and \
each row represents a gene. The colored cells in the plot at the intersection of each row & column represents the relative expression level of the \
gene in the sample. An additional layer of processing clusters samples and genes by their similarity to each other. So, genes with similar expression \
patterns across the samples cluster together (visualized by the dendrogram along the left side of the plot) \
and samples with overall similarity cluster together (visualized by the dendrogram across the top of the plot). \
The coloring represents relative expression level. Red is high expression & blue is low expression.")

select <- order(rowMeans(counts(dds, normalized=T)), decreasing=T)[1:1000]

pheatmap(assay(rldds)[select,], cluster_rows=T, cluster_cols=T, show_rownames=F, annotation_col=condition)

# tiff file
pheatmap(assay(rldds)[select,], cluster_rows=T, cluster_cols=T, show_rownames=F, annotation_col=condition, filename=paste("$prefix", "_Heatmap.tiff"))
```
[tiff heatmap](./DESeq2_Heatmap.tiff)

## Do pathway analysis

```{r pathway analysis}
BGD.05 <- as.character(dplyr::select(dplyr::filter(dplyr::arrange(res, padj), padj < 0.05), BestGeneDescriptor)$$BestGeneDescriptor)
rslt <- tryCatch(
        expr = {
            gprofile_Ordered <- gost(BGD.05, organism="$gProfilerkey", ordered_query=T, correction_method='gSCS', significant=T)
        },
        error = function(e){
            message("call to gost() generated an error")
            return(e)
        },
        warning = function(w){
            message("call to gost() generated a warning")
            return(w)
        },
        finally = {
            message("moving on")
        }
)
#
if (! inherits(rslt, 'simpleError')) {
# what we want to output contains a list, so collapse it
    gprofile_Ordered$$result$$parents <- paste(gprofile_Ordered$$result$$parents, collapse=',')
# write the gprofiler output to a file
    write.table(gprofile_Ordered['result'], file=paste0("$prefix","_gProfileR.txt"), sep="\t", quote=F, row.name=F, col.name=T)
} else {
    message("Since the pathway analysis generated an error, do not create output file.")
}
```

To do a pathway analysis, all the genes that satisfy a certain significance threshold are collected and subjected to g:GOSt Gene Group Functional
Profiling as implemented by [g:Profiler](https://academic.oup.com/nar/article/44/W1/W83/2499292). Typically, the threshold to include genes is that the
[DESeq2](https://bioconductor.org/packages/release/bioc/html/DESeq2.html) FDR-corrected p-value < 0.05.
All the tests are run using R on the command-line to have access to the latest algorithms and
databases. The default output of these efforts is a spreadsheet of data that includes the results of looking for pathways with a disproportionately
large number of genes from this experiment. The databases used in this analysis are:


* [KEGG](https://academic.oup.com/nar/article/42/D1/D199/1047899)
* [Reactome](https://academic.oup.com/nar/article/44/D1/D481/2503122)
* GO - [Gene Ontology](https://academic.oup.com/nar/article/43/D1/D1049/2439067)
* HP - [Human Phenotype Ontology](https://academic.oup.com/nar/article/42/D1/D966/1042793)
* COR - [CORUM](https://academic.oup.com/nar/article/38/suppl_1/D497/3112260) the comprehensive resource of mammalian protein complexes
* TF - [TRANSFAC](https://academic.oup.com/bib/article/9/4/326/266917) database of TRANS-acting FACtors


Each of these have been split out to their own tab in the following spreadsheet:

[pathway analysis results (tab-delimited text)](./DESeq2_gProfileR.txt)


## Save R data image file and session info
```{r save data image, include=FALSE}
save.image(file=paste0("$prefix", "_RData"))
```

```{r save sessionInfo, echo=FALSE}
sessionInfo()
```

